doc/zk-integration.md

```diff
@@ -0,0 +1,264 @@
+  block in the range. Note that such list of proofs could be aggregated or verified in batch depending on the commitment
+  scheme used to represent each Espresso block.
+- The _COMMs Equivalence_ gadget checks that using the same rollup inputs _ROLLUP_TXS_, we obtain _cm_txs_history_ using
+  the Espresso commitment scheme for representing a set of transactions and the commitment _cm_txs_rollup_ that is used
+  the by _zkVM_ gadget. This gadget is required in order to ensure that the set of transactions fetched from the
+  Espresso blocks and represented as _cm_txs_history_ is consistent with the set of transactions applied to the rollup
+  state and represented by the commitment _cm_txs_rollup_. Note that if both commitment schemes (used in Espresso and
+  the Rollup) were the same, this gadget would not be necessary. Thus, if updating the zkVM circuit is possible in
+  practice, by using the Espresso commitment scheme inside the zkVM gadget, one can remove the need for the _COMMs
+  Equivalence_ gadget.
+- The _zkVM_ gadget is the original gadget of the rollup circuit that proves a correct transition from state
+  _cm_state_vm i_ to the next state _cm_state_vm i+1_ when applying the transactions represented by the commitment value
+  _COMM_TXS_ROLLUP_.
+- These three gadgets above return a boolean: true if the verification succeeds and false otherwise.
+- For the circuit to accept, all these gadget outputs must be true, and thus we add an _AND_ gate.
+
+The pseudocode of the rollup contract below shows that in the case we rely on the Espresso light client contract to
+fetch the Espresso state, the only inputs to the function `updateRollupState` are `newVMState`, `commTxsRollup` and
+`snarkProof`.
+
+```solidity
+/// Uses the Espresso light client contract to fetch the last state.
+contract RollupContract1 is RollupContract {
+
+    function updateRollupState(
+        newVMState,
+        commTxsRollup,
+        snarkProof){
+
+        // Escape hatch is activated, switch to default sequencing mode
+        if (isEscapeHatchActivated()){
+            this.updateStateBackupSequencingMode(commTxsRollup);
+        } else { // No escape hatch, use the state of Espresso consensus
+            lightClientState = lcContract.getFinalizedState();
+            newEspressoState = lightClientState.blockCommRoot;
+            blockNumberEspresso = lightClientState.blockHeight;
+            this.updateStateFromEspresso(newEspressoState, blockNumberEspresso, commTxsRollup, newVMState, snarkProof);
+        }
+    }
+}
+```
+
+## Integration 2: Verify Espresso consensus inside the rollup circuit
+
+![image](zk-rollup-circuit.svg) **Figure 3:** Rollup circuit with additional gadgets _Espresso Consensus_, _Collect &
+Filter_, and _COMMs Equivalence_. Private inputs of the circuit are written in bold font.
+
+The circuit depicted in Figure 3 operates as follows:
```

Comment:
Seems that this circuit uses Figure 2 as a black box, then adds only _Espresso Consensus_ gadget. In that case, much of the following description is unnecessary.

Reply:
Right, simplification done in 6b93aa6b.